### PR TITLE
feat(datasets): add synthetic data generation pipeline

### PR DESCRIPTION
**Pull request by akotyla**

### CODE REVIEW

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+def get_closest_substring(long, short):
+    a, b = max(
+        combinations(re.finditer('|'.join(short.split()), long), 2),
+        key=lambda c: SequenceMatcher(None, long[c[0].start():c[1].end()], short).ratio()
+    )
+    return long[a.start():b.end()]
+
+def get_passages_list(raw_passages: str) -> list[str]:
+    passages = raw_passages.split("[")[1]
```
Comment (kdziedzic68) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py:13]: what if llm is mistaken about intended format ? or the raw passage lists looks lie that:

"['increment x by one' , sample random number from [x, x+1, x+2]']" ?

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/answer_gen_task.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        Formats the input data for the text generation model by creating a `ChatFormat` object.
+
+        Args:
+            input: A dictionary containing "chunk" (text source) and "question" keys.
+
+        Returns:
+            The formatted chat object containing the inputs for text generation.
+        """
+        chat = BasicAnswerGenPrompt(BasicAnswerGenInput(chunk=input["chunk"], question=input["question"])).chat
```
Comment (kdziedzic68) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/answer_gen_task.py:29]: i don't think that we should stick to any hardcoded prompt class for a task - prompt class should imo be an instance attribute

File: examples/evaluation/dataset-generator/generate.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        query_gen_task >> answer_gen_task >> passages_gen_task
+
+    distiset = pipeline.run(
+        use_cache=False,
+        dataset=dataset
+    )
+    result = distiset["default"]["train"]
+    result = result.remove_columns(["distilabel_metadata", "model_name"])
+
+    result.push_to_hub(
```
Comment (kdziedzic68) [inline examples/evaluation/dataset-generator/generate.py:55]: we should be able to store the dataset locally

Comment (kdziedzic68) [reply examples/evaluation/dataset-generator/generate.py:55]: dataset itself as well as yaml file representing it, compatible with evaluation / optimization frameworks

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/answer_gen_task.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+from typing import Any
+
+from distilabel.steps.tasks import TextGeneration
+from ragbits.core.prompt.base import ChatFormat
+
+from ragbits.evaluate.dataset_generator.prompt_basic_answer_gen import BasicAnswerGenPrompt, BasicAnswerGenInput
+
+class AnswerGenTask(TextGeneration):
```
Comment (kdziedzic68) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/answer_gen_task.py:8]: i think that we can create abstract class for DistiLabel task - things like input_names, output_names, prompt_class, prompt_input class should be passed in constructor - the target class would only need to implement format_output method

File: examples/evaluation/dataset-generator/generate.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+from distilabel.llms import OpenAILLM
```
Comment (kdziedzic68) [inline examples/evaluation/dataset-generator/generate.py:1]: i guess that distilabel is a new dependency. if i'm right requirements need update

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/passages_gen_task.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+from typing import Any
```
Comment (kdziedzic68) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/passages_gen_task.py:1]: think that dataset_generator should have subpackages - like qa for now in the future maybe corpus, ner, etc.

File: examples/evaluation/dataset-generator/generate.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        providers={DocumentType.PDF: UnstructuredPdfProvider(chunking_kwargs={"max_characters": 512})}
+    )
+    document_meta = DocumentMeta.from_local_path(local_path=CORPUS_PATH)
+    document_processor = router.get_provider(document_meta)
+    elements = await document_processor.process(document_meta)
+    dataset = Dataset.from_dict({
+        "chunk": [node.content for node in elements if node.element_type=="text"][:2]
+    })
+
+    with Pipeline("synthetic-RAG-data") as pipeline:
```
Comment (kdziedzic68) [inline examples/evaluation/dataset-generator/generate.py:31]: we should be able to stack configurable list of tasks into pipeline

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    Returns:
+        list of parsed passages
+    """
+    match = re.search(r"\[(.*?)\]", raw_passages, re.DOTALL)
+
+    if match:
+        passages_content = match.group(1)
+        try:
+            # Use eval to convert the string to a list, assuming it's a valid list-like format
+            return ast.literal_eval("[" + passages_content + "]")
```
Comment (pawel-chmielak-deepsense) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py:38]: You shouldn't have to use `eval` for that, especially on content returned by an LLM.

File: examples/evaluation/dataset-generator/config/generate.yaml
Comment (pawel-chmielak-deepsense) [inline examples/evaluation/dataset-generator/config/generate.yaml:1]: Is there a schema definition for these YAMLs somewhere?

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+from ragbits.core.utils.config_handling import get_cls_from_config
+
+module = sys.modules[__name__]
+
+
+class DatasetGenerationPipeline:
+    """A pipeline for dataset generation"""
+
+    def __init__(self, config: DictConfig):
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/pipeline.py:17]: afair there was an assumption that configuration of pipelien steps has to be generic here without assumptions of single tasks arguments, however agree thtat from the way how we use it there are at last SOME so i would handle it

[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/pipeline.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        """
+        Generates a dataset from a corpus or list of topics
+        Args:
+            corpus: a corpus of information or list of topics
+        Returns:
+            dataset instance
+        """
+        dataset = Dataset.from_dict({self.config.input_name: corpus})
+        with Pipeline(self.config.pipeline.name) as pipeline:
+            tasks = self._parse_pipeline_steps()
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/pipeline.py]: anyway - done

File: examples/evaluation/dataset-generator/generate.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+from ragbits.evaluate.dataset_generator.pipeline import DatasetGenerationPipeline
+from ragbits.evaluate.utils import log_dataset_to_file
+
+
+@hydra.main(config_path="config", config_name="generate", version_base="3.2")
+def main(config: DictConfig) -> None:
+    """
+    A main function for dataset generation example
+    Args:
+        config - configuration
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply examples/evaluation/dataset-generator/generate.py:13]: config is now validated through `pydantic.BaseModel` - all parameters are described within its definition

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/corpus_generation.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+module = sys.modules[__name__]
+
+
+class CorpusGenerationStep(Step):
+    """A step for corpus generation on given topics"""
+
+    def __init__(
+        self,
+        llm: LLM,
+        num_per_query: int,
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/corpus_generation.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/filter/dont_know.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    """A class for basic rule-based filtering of don't know anwers"""
+
+    def process(self, *inputs: StepInput) -> "StepOutput":
+        """
+        Runs the processing step
+        Args:
+            inputs - inputs to the step
+        Returns:
+            filtered outputs
+        """
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/filter/dont_know.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/filter/dont_know.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        result = [
+            {input_type: inp[input_type] for input_type in inp} for inp in inputs[0] if not self._is_dont_know(inp)
+        ]
+        yield result
+
+    @staticmethod
+    def _is_dont_know(inp: dict[str, Any]) -> bool:
+        return any(s.lower() in inp["basic_answer"].lower() for s in DONT_KNOW_PHRASES)
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/filter/dont_know.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/text_generation/qa.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        return {"chunk": input["chunk"], "question": output}  # type: ignore
+
+
+class PassagesGenTask(BaseDistilabelTask):
+    """
+    A task for generating passages related to a specific question and answer from a text chunk.
+    """
+
+    get_matches: bool = False
```
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/tasks/text_generation/qa.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py (hunk 8 above)
Comment (kdziedzic68) [reply packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py]: done

File: packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        raw_passages: string representing raw passages returned by llm
+    Returns:
+        list of parsed passages
+    """
+    match = re.search(r"\[(.*?)\]", raw_passages, re.DOTALL)
+
+    if match:
+        passages_content = match.group(1)
+        try:
+            # Use eval to convert the string to a list, assuming it's a valid list-like format
```
Comment (pawel-chmielak-deepsense) [inline packages/ragbits-evaluate/src/ragbits/evaluate/dataset_generator/utils.py:37]: Irrelevant comment

**Review by pawel-chmielak-deepsense — APPROVED** [review]

**Review by kdziedzic68 — APPROVED** [review]

**Pull request merged**